proof/crefine/RISCV64/Recycle_C.thy

```diff
@@ -805,16 +805,123 @@ lemma ntfn_q_refs'_no_NTFNBound[simp]:
+                      split: if_split_asm option.split)
+     apply ceqv
+    apply (rule ccorres_cond[where R=\<top>])
+      apply (clarsimp split: option.splits fault.splits)
+      apply (rename_tac fault s s', case_tac fault; clarsimp simp: seL4_Faults)
+      apply (rename_tac arch_fault, case_tac arch_fault, clarsimp simp: seL4_Arch_Faults)
+     apply (rule ccorres_rhs_assoc)+
+     apply (ctac add: setThreadState_ccorres)
+       apply (clarsimp simp: ifCondRefillUnblockCheck_def)
+       apply (rule ccorres_pre_threadGet)
+       apply (rule ccorres_move_c_guard_tcb)
+       apply (clarsimp simp: when_def)
+       apply (simp add: if_to_top_of_bind)
+       apply (rule ccorres_if_lhs)
+        apply (clarsimp simp: bind_assoc, rename_tac scPtr)
+        apply (wpfix add: option.sel)
+        apply (rule ccorres_pre_getObject_sc, rename_tac sc)
+        apply (rule ccorres_pre_getCurSc, rename_tac cur_sc)
+        apply (rule_tac xf'="ret__unsigned_long_'"
+                    and val="from_bool (scSporadic sc)"
+                    and R="obj_at' (\<lambda>tcb. tcbSchedContext tcb = Some scPtr) t and ko_at' sc scPtr
+                           and no_0_obj'"
+                    and R'=UNIV
+                     in ccorres_symb_exec_r_known_rv)
+           apply (rule conseqPre, vcg)
+           apply normalise_obj_at'
+           apply (frule (1) obj_at_cslift_tcb)
+           apply (frule (1) obj_at_cslift_sc)
+           apply (clarsimp simp: typ_heap_simps ctcb_relation_def csched_context_relation_def
+                                 to_bool_def
+                          split: if_splits)
+          apply ceqv
+         apply (rule ccorres_split_nothrow[where r'=dc and xf'=xfdc])
+             apply (rule ccorres_cond[where R=\<top>])
+               apply fastforce
+              apply (rule ccorres_move_c_guard_tcb)
+              apply (rule_tac Q="\<lambda>s. cur_sc = ksCurSc s
+                                     \<and> obj_at' (\<lambda>tcb. tcbSchedContext tcb = Some scPtr) t s"
+                           in ccorres_cond_both'[where Q'=\<top>])
+                apply clarsimp
+                apply (frule (1) obj_at_cslift_tcb)
+                apply (fastforce dest!: rf_sr_ksCurSC simp: typ_heap_simps' ctcb_relation_def)
+               apply (rule ccorres_move_c_guard_tcb)
+               apply (ctac add: refill_unblock_check_ccorres)
+              apply (rule ccorres_return_Skip)
+             apply (rule ccorres_return_Skip)
+            apply ceqv
+           apply (ctac add: possibleSwitchTo_ccorres)
+          apply wpsimp
+         apply (vcg exspec=refill_unblock_check_modifies)
+        apply (vcg exspec=sc_sporadic_modifies)
+       apply (rule_tac xf'="ret__unsigned_long_'"
+                   and val="from_bool False"
+                   and R="obj_at' (\<lambda>tcb. tcbSchedContext tcb = None) t"
+                   and R'=UNIV
+                    in ccorres_symb_exec_r_known_rv)
```

Comment:
trivial, but you can make this a bit shorter by writing `ccorres_symb_exec_r_known_rv[where R'=UNIV]` (same above), and to indicate that the R' isn't interesting